daffodil-runtime2/src/main/resources/org/apache/daffodil/runtime2/c/libruntime/errors.c

```diff
@@ -17,34 +17,109 @@
 
 #include "errors.h"
 #include <assert.h>    // for assert
-#include <error.h>     // for error
 #include <inttypes.h>  // for PRId64
 #include <stdbool.h>   // for bool, false, true
-#include <stdio.h>     // for NULL, feof, ferror, FILE, size_t
-#include <stdlib.h>    // for EXIT_FAILURE
+#include <stdio.h>     // for fprintf, stderr, NULL, feof, ferror, FILE, size_t, stdout
+#include <stdlib.h>    // for exit, EXIT_FAILURE
 
-// error_message - return an internationalized error message
+// eof_or_error - get pointer to error if stream has eof or error indicator set
+
+const Error *
+eof_or_error(FILE *stream)
+{
+    if (feof(stream))
+    {
+        static Error error = {ERR_STREAM_EOF, {NULL}};
+        return &error;
+    }
+    else if (ferror(stream))
+    {
+        static Error error = {ERR_STREAM_ERROR, {NULL}};
+        return &error;
+    }
+    else
+    {
+        return NULL;
+    }
+}
+
+// get_diagnostics - get pointer to validation diagnostics
+
+Diagnostics *
+get_diagnostics(void)
+{
+    static Diagnostics diagnostics;
+    return &diagnostics;
+}
+
+// add_diagnostic - add a new error to validation diagnostics
+
+bool
+add_diagnostic(Diagnostics *diagnostics, const Error *error)
+{
+    if (diagnostics && error)
+    {
+        if (diagnostics->length < LIMIT_DIAGNOSTICS)
+        {
+            Error *err = &diagnostics->array[diagnostics->length++];
+            err->code = error->code;
+            err->s = error->s;
+            return true;
+        }
+    }
+    return false;
+}
+
+// error_message - get an internationalized error message
 
 static const char *
 error_message(enum ErrorCode code)
```

Comment:
Feels like this should be in a CLI-oriented file, since it's giving CLI-help.

Reply:
Yeah, I originally had these CLI messages in daffodil_getopt.c and thought upon it for a while before deciding it was easier to bundle all the messages in one place for easier internationalization and flow of control (daffodil_parse_cli needs only return statements, not printf and exit calls too).  However, if we don't care about hardcoding 'daffodil' in the help, daffodil_parse_cli could pass the help message in err->s and both CLI_HELP_USAGE and CLI_PROGRAM_VERSION then would use the same format, "%s\n".  Would you prefer that?

Reply:
To me, feels like anything CLI related should not be in libruntime. The CLI is just a user of libruntime, and just like any other runtime user, they shouldn't need to update the runtime to change user specific error messages. Almost feels like the CLI needs a completely separate error path, or libruntime needs someway for users to inject addition error capabilities.

Reply:
Sounds like you would prefer that libcli have its own cli_errors.[ch] with similar enum, struct, and functions just to keep all the CLI_* codes and messages outside of libruntime.  Mike, do you want that too?

Reply:
That makes sense to me yes. We don't have to do this for this change set. We can do it subsequently if you prefer. But we definitely want libruntime to be usable from an application which has nothing whatsoever to do with CLIs nor XML,  and has its own error architecture used to create error messages from the error-codes coming back from the libruntime.